### PR TITLE
add hid_raw feature to VUSB

### PR DESCRIPTION
**Pull request by hsgw**
## Description

Add hid_raw feature to keyboard with USB. 

Currently,  I had tested on atmega328p.   
Would anyone test on other VUSB MCU?
Test method is here.  https://github.com/qmk/qmk_firmware/issues/8320#issuecomment-597762268

### Limitation
This feature is not available with  mousekey / extrakey.

## TODO
- [ ] Add documentation for setting in `usbconfig.h`

~I will remove draft flag after #8321 merged.~

## Types of Changes

- [x] Core
- [ ] Bugfix
- [x] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* #8320

## Checklist

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [ ] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: tmk_core/protocol/vusb/vusb.c [hunk 1]
```diff
@@ -28,6 +28,14 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
 #include "vusb.h"
 #include <util/delay.h>
 
+#ifdef RAW_ENABLE
+#    include "raw_hid.h"
+#endif
+
+#if (defined(MOUSE_ENABLE) || defined(EXTRAKEY_ENABLE)) && defined(RAW_ENABLE)
+#    error "can't use mouse/extra key and hid raw at same time on VUSB"
+#endif
```
Comment (drashna) [inline tmk_core/protocol/vusb/vusb.c]: IIRC, this is handled automatically, and should error out anyways, if there are too many endpoints enabled.

Comment (hsgw) [reply tmk_core/protocol/vusb/vusb.c]: Should this be handled in makefile? Now this is not implemented in makefile.

Comment (Wilba6582) [reply tmk_core/protocol/vusb/vusb.c]: This test is good as it is now, as it's explicit about the mutual exclusivity and conditional compilation requirement of this code.

Comment (hsgw) [reply tmk_core/protocol/vusb/vusb.c]: Thanks, Leave this as it is.

File: tmk_core/protocol/vusb/main.c [hunk 2]
```diff
@@ -6,35 +6,38 @@
...
 #include "keyboard.h"
 #include "host.h"
 #include "timer.h"
 #include "uart.h"
 #include "debug.h"
 
-#if defined(RGBLIGHT_ENABLE)
+#if (defined(RGB_MIDI) || defined(RGBLIGHT_ANIMATIONS)) && defined(RGBLIGHT_ENABLE)
 #    include "rgblight.h"
 #endif
```
Comment (mtei) [inline tmk_core/protocol/vusb/main.c]:
```suggestion
```
In #7733, the call to `rgblight_task()` was removed from `protocol/vusb/main.c`, so this import should no longer be necessary.

**Comment by yiancar** [comment]
Confirmed at working on Gingham as well. If anyone feels that the comments are not enough please let us know and we will add more of them:)

File: tmk_core/protocol/vusb/vusb.c (hunk 1 above)
Comment (drashna) [reply tmk_core/protocol/vusb/vusb.c]: Ah, okay, LUFA and ChibiOS has a check for the number of endpoints, but V-USB doesn't.  

This is fine, for now, but is somethign that we would want to change in the near future, @fauxpark?

**Comment by coseyfannitutti** [comment]
@hsgw webhid test confirmed with atmega32a

**Comment by hsgw** [comment]
I did not notice conflicts, I'll fix tonight.

**Comment by hsgw** [comment]
I have rewritten the descriptors according to the new style.

**Comment by yiancar** [comment]
As we are not going to make vusb a submodule until may could we please merge this now?:)

File: tmk_core/protocol/vusb/vusb.c [hunk 3]
```diff
@@ -572,6 +720,12 @@ USB_PUBLIC usbMsgLen_t usbFunctionDescriptor(struct usbRequest *rq) {
                     usbMsgPtr = (unsigned char *)&usbConfigurationDescriptor.mouseExtraHID;
                     len       = sizeof(usbHIDDescriptor_t);
                     break;
+#endif
+#if defined(RAW_ENABLE)
```
Comment (fauxpark) [inline tmk_core/protocol/vusb/vusb.c]:
```suggestion
#elif defined(RAW_ENABLE)
```

File: tmk_core/protocol/vusb/vusb.c [hunk 4]
```diff
@@ -587,6 +741,12 @@ USB_PUBLIC usbMsgLen_t usbFunctionDescriptor(struct usbRequest *rq) {
                     usbMsgPtr = (unsigned char *)mouse_extra_hid_report;
                     len       = sizeof(mouse_extra_hid_report);
                     break;
+#endif
+#if defined(RAW_ENABLE)
```
Comment (fauxpark) [inline tmk_core/protocol/vusb/vusb.c]:
```suggestion
#elif defined(RAW_ENABLE)
```

File: tmk_core/protocol/vusb/vusb.h [hunk 5]
```diff
@@ -98,9 +98,22 @@ typedef struct usbConfigurationDescriptor {
     usbEndpointDescriptor_t mouseExtraINEndpoint;
 #    endif
 #endif
+
+#if defined(RAW_ENABLE)
```
Comment (fauxpark) [inline tmk_core/protocol/vusb/vusb.h]:
```suggestion
#elif defined(RAW_ENABLE)
```

File: tmk_core/protocol/vusb/vusb.c [hunk 6]
```diff
@@ -572,6 +720,12 @@ USB_PUBLIC usbMsgLen_t usbFunctionDescriptor(struct usbRequest *rq) {
                     usbMsgPtr = (unsigned char *)&usbConfigurationDescriptor.mouseExtraHID;
                     len       = sizeof(usbHIDDescriptor_t);
                     break;
+#endif
+#if defined(RAW_ENABLE)
+                case 1:
+                    usbMsgPtr = (unsigned char *)&usbConfigurationDescriptor.rawHID;
+                    len       = 9;
```
Comment (fauxpark) [inline tmk_core/protocol/vusb/vusb.c]:
```suggestion
                    len       = sizeof(usbHIDDescriptor_t);
```

File: tmk_core/protocol/vusb/vusb.c [hunk 7]
```diff
@@ -512,6 +612,54 @@ const PROGMEM usbConfigurationDescriptor_t usbConfigurationDescriptor = {
...
+            .bLength         = sizeof(usbHIDDescriptor_t),
+            .bDescriptorType = USBDESCR_HID
+        },
+        .bcdHID              = 0x0101,
+        .bCountryCode        = 0x00,
+        .bNumDescriptors     = 2,
+        .bDescriptorType     = USBDESCR_HID_REPORT,
+        .wDescriptorLength   = sizeof(raw_hid_report)
+    },
+#        if USB_CFG_HAVE_INTRIN_ENDPOINT3 /* endpoint descriptor for endpoint 3 */
```
Comment (fauxpark) [inline tmk_core/protocol/vusb/vusb.c]:
```suggestion
#        if USB_CFG_HAVE_INTRIN_ENDPOINT3
```

File: tmk_core/protocol/vusb/vusb.c [hunk 8]
```diff
@@ -512,6 +612,54 @@ const PROGMEM usbConfigurationDescriptor_t usbConfigurationDescriptor = {
     }
 #        endif
 #    endif
+#    if defined(RAW_ENABLE)
+    .rawInterface  = {
+        .header = {
+            .bLength = sizeof(usbInterfaceDescriptor_t),
```
Comment (fauxpark) [inline tmk_core/protocol/vusb/vusb.c]:
```suggestion
            .bLength         = sizeof(usbInterfaceDescriptor_t),
```

File: tmk_core/protocol/vusb/vusb.c [hunk 9]
```diff
@@ -512,6 +612,54 @@ const PROGMEM usbConfigurationDescriptor_t usbConfigurationDescriptor = {
     }
 #        endif
 #    endif
+#    if defined(RAW_ENABLE)
```
Comment (fauxpark) [inline tmk_core/protocol/vusb/vusb.c]:
```suggestion
#    elif defined(RAW_ENABLE)
```

File: tmk_core/protocol/vusb/vusb.c [hunk 10]
```diff
@@ -512,6 +612,54 @@ const PROGMEM usbConfigurationDescriptor_t usbConfigurationDescriptor = {
...
+    .rawInterface  = {
```
Comment (fauxpark) [inline tmk_core/protocol/vusb/vusb.c]:
```suggestion
    .rawInterface = {
```

File: tmk_core/protocol/vusb/vusb.c [hunk 11]
```diff
@@ -335,6 +410,31 @@ const PROGMEM uchar mouse_extra_hid_report[] = {
...
+                           //   (0 << 0) | (1 << 1) | (0 << 2)
+    // Data from host
+    0x09, 0x63,             //   Vender Defined
+    0x15, 0x00,             //   Logical Minimum
+    0x26, 0xFF, 0x00,       //   Logical Maximum
+    0x95, RAW_OUTPUT_SIZE,  //   Report Count (32)
+    0x75, 0x08,             //   Report Size (8)
+    0x91, 0x02,             //   Output (Data, Variable, Absolute, None-Volatile)
+                            //   (0 << 0) | (1 << 1) | (0 << 2) | (0 << 7)
+    0xC0,                   // End Collection
```
Comment (fauxpark) [inline tmk_core/protocol/vusb/vusb.c]:
```suggestion
    0x06, 0x60, 0xFF,  // Usage Page (Vendor Defined)
    0x09, 0x61,        // Usage (Vendor Defined)
    0xA1, 0x01,        // Collection (Application)
    // Data to host
    0x09, 0x62,            //   Usage (Vendor Defined)
    0x15, 0x00,            //   Logical Minimum (0)
    0x26, 0xFF, 0x00,      //   Logical Maximum (255)
    0x95, RAW_INPUT_SIZE,  //   Report Count
    0x75, 0x08,            //   Report Size (8)
    0x81, 0x02,            //   Input (Data, Variable, Absolute)
    // Data from host
    0x09, 0x63,             //   Usage (Vendor Defined)
    0x15, 0x00,             //   Logical Minimum (0)
    0x26, 0xFF, 0x00,       //   Logical Maximum (255)
    0x95, RAW_OUTPUT_SIZE,  //   Report Count
    0x75, 0x08,             //   Report Size (8)
    0x91, 0x02,             //   Output (Data, Variable, Absolute)
    0xC0,                   // End Collection
```

File: tmk_core/protocol/vusb/vusb.c [hunk 12]
```diff
@@ -71,6 +79,52 @@ void vusb_transfer_keyboard(void) {
...
 
+/*------------------------------------------------------------------*
+ * RAW HID
+ *------------------------------------------------------------------*/
+#ifdef RAW_ENABLE
+#    define RAW_INPUT_SIZE (32)
+#    define RAW_OUTPUT_SIZE (32)
+
+static uint8_t raw_output_buffer[RAW_OUTPUT_SIZE];
+static uint8_t raw_output_recieved_bytes = 0;
```
Comment (fauxpark) [inline tmk_core/protocol/vusb/vusb.c]:
```suggestion
static uint8_t raw_output_received_bytes = 0;
```
;)

Comment (hsgw) [reply tmk_core/protocol/vusb/vusb.c]: I missed this!
But this is corrected
Thanks a lot :)

Comment (fauxpark) [reply tmk_core/protocol/vusb/vusb.c]: I meant that this is actually the correct spelling, and you should replace the rest of the instances of `raw_output_recieved_bytes`.

Comment (hsgw) [reply tmk_core/protocol/vusb/vusb.c]: I What I wanted to say is that this is already fixed.
I'm sorry for my poor English.

Comment (fauxpark) [reply tmk_core/protocol/vusb/vusb.c]: That commit changes it back to the wrong spelling.

Comment (hsgw) [reply tmk_core/protocol/vusb/vusb.c]: ahhhhh. I made misstake. thanks.

File: tmk_core/protocol/vusb/vusb.c [hunk 13]
```diff
@@ -28,6 +28,14 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
 #include "vusb.h"
 #include <util/delay.h>
 
+#if defined(RAW_ENABLE)
+#    include "raw_hid.h"
+#endif
+
+#if (defined(MOUSE_ENABLE) || defined(EXTRAKEY_ENABLE)) && defined(RAW_ENABLE)
+#    error "can't use mouse/extra key and hid raw at same time on VUSB"
```
Comment (fauxpark) [inline tmk_core/protocol/vusb/vusb.c]:
```suggestion
#    error "Enabling Mousekeys/Extrakeys and Raw HID at the same time is not currently supported on V-USB."
```

File: tmk_core/protocol/vusb/vusb.c [hunk 14]
```diff
@@ -71,6 +79,52 @@ void vusb_transfer_keyboard(void) {
     }
 }
 
+/*------------------------------------------------------------------*
+ * RAW HID
+ *------------------------------------------------------------------*/
+#ifdef RAW_ENABLE
+#    define RAW_INPUT_SIZE (32)
+#    define RAW_OUTPUT_SIZE (32)
```
Comment (fauxpark) [inline tmk_core/protocol/vusb/vusb.c]:
```suggestion
#    define RAW_EPSIZE 32
```
We really only need the one here.

File: tmk_core/protocol/vusb/vusb.c [hunk 15]
```diff
@@ -71,6 +79,52 @@ void vusb_transfer_keyboard(void) {
...
+
+static uint8_t raw_output_buffer[RAW_OUTPUT_SIZE];
```
Comment (fauxpark) [inline tmk_core/protocol/vusb/vusb.c]:
```suggestion
static uint8_t raw_output_buffer[RAW_EPSIZE];
```

File: tmk_core/protocol/vusb/vusb.c [hunk 16]
```diff
@@ -71,6 +79,52 @@ void vusb_transfer_keyboard(void) {
...
+static uint8_t raw_output_received_bytes = 0;
+
+void raw_hid_send(uint8_t *data, uint8_t length) {
+    if (length != RAW_INPUT_SIZE) {
```
Comment (fauxpark) [inline tmk_core/protocol/vusb/vusb.c]:
```suggestion
    if (length != RAW_EPSIZE) {
```

File: tmk_core/protocol/vusb/vusb.c [hunk 17]
```diff
@@ -71,6 +79,52 @@ void vusb_transfer_keyboard(void) {
...
+
+__attribute__((weak)) void raw_hid_receive(uint8_t *data, uint8_t length) {
+    // Users should #include "raw_hid.h" in their own code
+    // and implement this function there. Leave this as weak linkage
+    // so users can opt to not handle data coming in.
+}
+
+void raw_hid_task(void) {
+    if (raw_output_received_bytes == RAW_OUTPUT_SIZE) {
+        raw_hid_receive(raw_output_buffer, RAW_OUTPUT_SIZE);
```
Comment (fauxpark) [inline tmk_core/protocol/vusb/vusb.c]:
```suggestion
    if (raw_output_received_bytes == RAW_EPSIZE) {
        raw_hid_receive(raw_output_buffer, RAW_EPSIZE);
```

File: tmk_core/protocol/vusb/vusb.c [hunk 18]
```diff
@@ -206,6 +260,27 @@ uchar usbFunctionWrite(uchar *data, uchar len) {
...
+void usbFunctionWriteOut(uchar *data, uchar len) {
+#ifdef RAW_ENABLE
+    // Data from host must be divided every 8bytes
+    if (len != 8) {
+        debug("RAW: invalid length");
+        raw_output_received_bytes = 0;
+        return;
+    }
+
+    if (raw_output_received_bytes + len > RAW_OUTPUT_SIZE) {
```
Comment (fauxpark) [inline tmk_core/protocol/vusb/vusb.c]:
```suggestion
    if (raw_output_received_bytes + len > RAW_EPSIZE) {
```

File: tmk_core/protocol/vusb/vusb.c [hunk 19]
```diff
@@ -335,6 +410,29 @@ const PROGMEM uchar mouse_extra_hid_report[] = {
...
+#if defined(RAW_ENABLE)
+const PROGMEM uchar raw_hid_report[] = {
+    0x06, 0x60, 0xFF,  // Usage Page (Vendor Defined)
+    0x09, 0x61,        // Usage (Vendor Defined)
+    0xA1, 0x01,        // Collection (Application)
+    // Data to host
+    0x09, 0x62,            //   Usage (Vendor Defined)
+    0x15, 0x00,            //   Logical Minimum (0)
+    0x26, 0xFF, 0x00,      //   Logical Maximum (255)
+    0x95, RAW_INPUT_SIZE,  //   Report Count
```
Comment (fauxpark) [inline tmk_core/protocol/vusb/vusb.c]:
```suggestion
    0x95, RAW_EPSIZE,      //   Report Count
```

File: tmk_core/protocol/vusb/vusb.c [hunk 20]
```diff
@@ -335,6 +410,29 @@ const PROGMEM uchar mouse_extra_hid_report[] = {
...
+    0x75, 0x08,            //   Report Size (8)
+    0x81, 0x02,            //   Input (Data, Variable, Absolute)
+    // Data from host
+    0x09, 0x63,             //   Usage (Vendor Defined)
+    0x15, 0x00,             //   Logical Minimum (0)
+    0x26, 0xFF, 0x00,       //   Logical Maximum (255)
+    0x95, RAW_OUTPUT_SIZE,  //   Report Count
```
Comment (fauxpark) [inline tmk_core/protocol/vusb/vusb.c]:
```suggestion
    0x95, RAW_EPSIZE,       //   Report Count
```

File: tmk_core/protocol/vusb/vusb.c [hunk 21]
```diff
@@ -511,6 +609,53 @@ const PROGMEM usbConfigurationDescriptor_t usbConfigurationDescriptor = {
...
+    },
+#        if USB_CFG_HAVE_INTRIN_ENDPOINT3
+    .rawINEndpoint = {
+        .header = {
+            .bLength         = sizeof(usbEndpointDescriptor_t),
+            .bDescriptorType = USBDESCR_ENDPOINT
+        },
+        .bEndpointAddress    = (USBRQ_DIR_DEVICE_TO_HOST | USB_CFG_EP3_NUMBER),
+        .bmAttributes        = 0x03,
+        .wMaxPacketSize      = 8,
```
Comment (fauxpark) [inline tmk_core/protocol/vusb/vusb.c]:
```suggestion
        .wMaxPacketSize      = RAW_EPSIZE,
```

File: tmk_core/protocol/vusb/vusb.c [hunk 22]
```diff
@@ -511,6 +609,53 @@ const PROGMEM usbConfigurationDescriptor_t usbConfigurationDescriptor = {
...
+        .bInterval           = USB_POLLING_INTERVAL_MS
+    },
+    .rawOUTEndpoint = {
+        .header = {
+            .bLength         = sizeof(usbEndpointDescriptor_t),
+            .bDescriptorType = USBDESCR_ENDPOINT
+        },
+        .bEndpointAddress    = (USBRQ_DIR_HOST_TO_DEVICE | USB_CFG_EP3_NUMBER),
+        .bmAttributes        = 0x03,
+        .wMaxPacketSize      = 8,
```
Comment (fauxpark) [inline tmk_core/protocol/vusb/vusb.c]:
```suggestion
        .wMaxPacketSize      = RAW_EPSIZE,
```

File: tmk_core/protocol/vusb/vusb.c (hunk 1 above)
Comment (fauxpark) [reply tmk_core/protocol/vusb/vusb.c]: @zvecr 's console PR #8559 includes something like that - perhaps we should pull that out and merge it first to simplify both these PRs.

File: tmk_core/protocol/vusb/vusb.c (hunk 14 above)
Comment (hsgw) [reply tmk_core/protocol/vusb/vusb.c]: EPSIZE is limited to "8" by VUSB.   
If we need simpler, what do you think of this?

```suggestion
#    define RAW_BUFFER_SIZE 32
#    define RAW_EPSIZE 8
```

Comment (fauxpark) [reply tmk_core/protocol/vusb/vusb.c]: Ah, this seems to be a limitation of USB 1.1 itself, not V-USB:
https://www.beyondlogic.org/usbnutshell/usb4.shtml#Interrupt
> The maximum data payload size for low-speed devices is 8 bytes.

So yes, `RAW_EPSIZE` should be 8. Not sure what the buffer size would be used for?

Comment (hsgw) [reply tmk_core/protocol/vusb/vusb.c]: > Ah, this seems to be a limitation of USB 1.1 itself, not V-USB:

But most hosts will be able to accept it.

`RAW_BUFFER_SIZE` is used to check the report length for compatibility with other protocols, as shown below.
```C
void raw_hid_send(uint8_t *data, uint8_t length) {
    if (length != RAW_BUFFER_SIZE) {
...
raw_hid_receive(raw_output_buffer, RAW_BUFFER_SIZE);
```

Comment (fauxpark) [reply tmk_core/protocol/vusb/vusb.c]: Right, so we would set the endpoint size to 8 and the report size to 32? That sounds fair enough.

Comment (hsgw) [reply tmk_core/protocol/vusb/vusb.c]: > we would set the endpoint size to 8 and the report size to 32?
Yes, I'll work on this tonight.

Comment (hsgw) [reply tmk_core/protocol/vusb/vusb.c]: I finished! Does this look good?

Comment (fauxpark) [reply tmk_core/protocol/vusb/vusb.c]: Yup!

**Comment by zvecr** [comment]
Thanks!

**Pull request merged**